### ISSUE TITLE
ChangePropertyKey not respecting sequentiality

### ISSUE THREAD

**Issue by JoseLora** [openrewrite/rewrite#3270]
Hi folks! I don't understand sequentiality in some OpenRewrite recipes. I reproduce this simple example:

```
public Sequential() {
    this.doNext(new ChangePropertyKey("common.tracing", "common.execution.context", true, null, null));
    this.doNext(new DeleteProperty("common.sleuth.send", null, null, null));
    this.doNext(new ChangePropertyKey("common.sleuth", "common.tracing", true, null, null));
  }
```
And this test should pass, but it fails because where I expect `common.tracing` appears as `common.execution.context`. If the changes are applied sequentially, it has no sense!

```
void changeOnlyDesiredProperties() {
    this.rewriteRun(yaml("""
        common:
          cache:
            enabled: true
          sleuth:
            send: true
            sender:
              http:
                username: user
                password: pwd
                connectionRequestTimeout: 5000
                timeout: 3000
                readTimeout: 3000
          blockhound.enabled: true
        """,
        """
            common:
              cache:
                enabled: true
              tracing:
                sender:
                  http:
                    username: user
                    password: pwd
                    connectionRequestTimeout: 5000
                    timeout: 3000
                    readTimeout: 3000
              blockhound.enabled: true
                    """));
  }
```

[Slack Message](https://rewriteoss.slack.com/archives/C01A843MWG5/p1684859716339379?thread_ts=1684859716.339379&cid=C01A843MWG5)

**Comment by timtebeek** [comment]
We had a [similar fix just last week](https://github.com/openrewrite/rewrite/commit/0a6b55efdbe9214834ce0ea72e1f78eb14b56f95) for the Maven recipe `RenamePropertyKey`;
Perhaps a similar approach would work here for this Yaml recipe.

Would you be interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?

**Comment by RossZhang-cpu** [comment]
Hi guys@timtebeek @JoseLora , would you mind let me take on this issue? I am willing to spend some time to investigate this one.

**Comment by RossZhang-cpu** [comment]
> We had a [similar fix just last week](https://github.com/openrewrite/rewrite/commit/0a6b55efdbe9214834ce0ea72e1f78eb14b56f95) for the Maven recipe `RenamePropertyKey`; Perhaps a similar approach would work here for this Yaml recipe.
> 
> Would you be interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?

Hi @timtebeek, It looks like there isn't same functional method(maybeUpdateModel()) in YamlVisitor.class. I am thinking of having to implemet one same functionality method for YamlVisitor.  Just want to check with you whether my understanding is correct?

**Comment by timtebeek** [comment]
Hi @RossZhang-cpu ; I think that `maybeUpdateModel` is pretty unique to our `MavenVisitor`, and only needed there to [update a marker](https://github.com/openrewrite/rewrite/blob/28e577966fd1033799442b2fe1dbb30dbabbefed/rewrite-maven/src/main/java/org/openrewrite/maven/UpdateMavenModel.java#L30) containing information derived from the Maven pom.xml file contents. For Yaml our model is simpler, and does not have that concept of derived information markers that I know of, so no need to write an equivalent `maybeUpdateModel`.

In this case I suspect it's mostly an issue with the recipes themselves, but perhaps we can start with a draft PR containing just the above test, and explore what's needed with a debugger from there on out.

**Comment by RossZhang-cpu** [comment]
Thank you @timtebeek for classification. I raised a draft PR only containing test, please kindly help to check.

**Comment by RossZhang-cpu** [comment]
Hi @timtebeek, after debugging I found the defaut value of cycle is 2, which result into another Recipe Execution and change the property key again. Please see the snapshots below:
![image](https://github.com/openrewrite/rewrite/assets/35285514/3e576e8f-5665-4ab8-87d6-f1611045f531)

Here maxCycles is 2 will cause another unexpected Recipe execution.
![image](https://github.com/openrewrite/rewrite/assets/35285514/b2219448-f30f-42ea-ad76-2e36c7753d1b)

It seems can't customize cycles vale from the begining in current codes. Do you have any clue about it?

**Comment by knutwannheden** [comment]
Yes, the test framework always runs 2 cycles. The reason is that any recipe can request that another cycle should be run, so the test framework always runs 2 cycles to make sure that the recipe behaves correctly in this scenario.

Here I believe the recipe is actually behaving as expected as the third recipe produces a new `common.tracing` property which in the second cycle will be transformed by the first recipe.

To avoid this the simplest solution is to set maxCycles: 1` on the YAML recipe, which I believe would be adequate in this use case.

So in summary I believe there is nothing to be fixed for this particular issue. By creating a test case and stepping through with the debugger and finding what caused this behavior I think you learned a lot about OpenRewrite and hope you don't feel your time was completely wasted and can instead apply that knowledge to look into another issue that is bothering you.

Unless you or @timtebeek see something else that needs fixing here, I suggest that we close the issue and the PR again.

**Comment by RossZhang-cpu** [comment]
Thank you @knutwannheden  for sharing information.
Even debugger takes me sometime, I am still feel happy to learn something new during that time.

And I have tried to set cycle to be 1 in test code, it works fine. 

```java:
rewriteRun(
                spec -> {
                    spec.recipeFromYaml(yamlRecipe, "org.openrewrite.ChangePropertyKeySequentialityTest");
                    spec.cycles(1);
                    }
```


But may I know how to implement you mentioned before set on YAML recipe
```
To avoid this the simplest solution is to set maxCycles: 1` on the YAML recipe
```

**Comment by knutwannheden** [comment]
Since `Recipe` declares a `maxCycles()` method I was thinking we probably allow declarative YAML recipes to specify a corresponding value. But it turns out we don't. This could be an interesting addition for use cases as in this issue. Let's see what @timtebeek thinks about this.

**Comment by timtebeek** [comment]
So Knut's proposal is up for discussion in:
- https://github.com/openrewrite/rewrite/pull/3850

With that I think we've all learned something here and I think this issue can be closed. Thanks for taking a look @RossZhang-cpu ! Always good seeing how you dive into issues here.
